core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala

```diff
@@ -506,20 +593,53 @@ final class DefaultSource extends ColumnarRelationProvider {
       schema, externalStore)
 
     var success = false
-    val relation = new ColumnFormatRelation(SnappyStoreHiveCatalog.
+
+    // create an index relation if it is an index table
+    val relation = options.get(StoreUtils.GEM_INDEXED_TABLE) match {
+      case Some(baseTable) => new IndexColumnFormatRelation(SnappyStoreHiveCatalog.
+        processTableIdentifier(table, sqlContext.conf),
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext,
+        baseTable)
+      case None => new ColumnFormatRelation(SnappyStoreHiveCatalog.
         processTableIdentifier(table, sqlContext.conf),
-      getClass.getCanonicalName,
-      mode,
-      schema,
-      schemaExtension,
-      ddlExtensionForShadowTable,
-      options,
-      externalStore,
-      blockMap,
-      partitioningColumn,
-      sqlContext)
+        getClass.getCanonicalName,
+        mode,
+        schema,
+        schemaExtension,
+        ddlExtensionForShadowTable,
+        options,
+        externalStore,
+        blockMap,
+        partitioningColumn,
+        sqlContext)
+    }
+
     try {
       relation.createTable(mode)
+      var indexes: Array[String] = Array()
+      try {
+        val options = new CaseInsensitiveMutableHashMap(relation.origOptions)
+        indexes = options(ExternalStoreUtils.INDEX_NAME).split(",")
+      } catch {
+        case e: NoSuchElementException =>
+        }
+
+      indexes.foreach(index => {
+        val sncCatalog = sqlContext.asInstanceOf[SnappyContext].catalog
+        val dr = sncCatalog.lookupRelation(sncCatalog.newQualifiedTableName(index)) match {
+          case LogicalRelation(r: DependentRelation, _ ) => r
+        }
+        relation.addDependent(dr, sncCatalog)
+      })
```

Comment:
Lets make this a generic thing. Whenever a DependentRelation is created, then it needs to be registered against its Parent, so this can be handled by a generic method somewhere. We could even add a separate class like "DependencyManager" or else let DependencyCatalog take care of such.